File: components/signaling-protocol/src/main/kotlin/com/battleship/shared/Result.kt
package com.battleship.shared

sealed class Result<out S, out F> {

    data class Success<S>(val value: S) : Result<S, Nothing>()
    data class Failure<F>(val reason: F) : Result<Nothing, F>()

    fun <T> map(transform: (S) -> T): Result<T, F> = when (this) {
        is Success -> Success(transform(value))
        is Failure -> this
    }

    fun <T> flatMap(transform: (@UnsafeVariance S) -> Result<T, @UnsafeVariance F>): Result<T, F> = when (this) {
        is Success -> transform(value)
        is Failure -> this
    }

    fun <F2> or(recover: (F) -> Result<@UnsafeVariance S, F2>): Result<S, F2> = when (this) {
        is Success -> this
        is Failure -> recover(reason)
    }

    fun tee(action: (@UnsafeVariance S) -> Unit): Result<S, F> = when (this) {
        is Success -> also { action(value) }
        is Failure -> this
    }

    fun <T> either(onSuccess: (S) -> T, onFailure: (F) -> T): T = when (this) {
        is Success -> onSuccess(value)
        is Failure -> onFailure(reason)
    }
}

fun <S> S.asSuccess(): Result<S, Nothing> = Result.Success(this)
fun <F> F.asFailure(): Result<Nothing, F> = Result.Failure(this)

inline fun <S, F> tryCatch(fn: () -> S, onError: (Throwable) -> F): Result<S, F> =
    try { fn().asSuccess() }
    catch (e: Throwable) { onError(e).asFailure() }

File: applications/web-client/src/lib/result.ts
export type Result<S, F> = Success<S, F> | Failure<S, F>

export type Success<S, F> = {
    readonly kind: 'success'
    readonly value: S
    map<T>(fn: (value: S) => T): Success<T, F>
    andThen<T, F2>(fn: (value: S) => Result<T, F2>): Result<T, F2>
    or<F2>(fn: (reason: F) => Result<S, F2>): Success<S, F2>
    onSuccess(fn: (value: S) => void): Success<S, F>
    onFailure(fn: (reason: F) => void): Success<S, F>
    either<S2, F2>(onSuccess: (value: S) => Result<S2, F2>, onFailure: (reason: F) => Result<S2, F2>): Result<S2, F2>
    mapEither<T>(onSuccess: (value: S) => T, onFailure: (reason: F) => T): T
}

export type Failure<S, F> = {
    readonly kind: 'failure'
    readonly reason: F
    map<T>(fn: (value: S) => T): Failure<T, F>
    andThen<T, F2>(fn: (value: S) => Result<T, F2>): Failure<T, F>
    or<F2>(fn: (reason: F) => Result<S, F2>): Result<S, F2>
    onSuccess(fn: (value: S) => void): Failure<S, F>
    onFailure(fn: (reason: F) => void): Failure<S, F>
    either<S2, F2>(onSuccess: (value: S) => Result<S2, F2>, onFailure: (reason: F) => Result<S2, F2>): Result<S2, F2>
    mapEither<T>(onSuccess: (value: S) => T, onFailure: (reason: F) => T): T
}

export const success = <S, F = never>(value: S): Success<S, F> => Object.freeze({
    kind: 'success' as const,
    value,
    map: (fn) => success(fn(value)),
    andThen: (fn) => fn(value),
    or: () => success(value),
    onSuccess: (fn) => { fn(value); return success(value) },
    onFailure: () => success(value),
    either: (onSuccess) => onSuccess(value),
    mapEither: (onSuccess) => onSuccess(value),
})

export const failure = <F, S = never>(reason: F): Failure<S, F> => Object.freeze({
    kind: 'failure' as const,
    reason,
    map: () => failure(reason),
    andThen: () => failure(reason),
    or: (fn) => fn(reason),
    onSuccess: () => failure(reason),
    onFailure: (fn) => { fn(reason); return failure(reason) },
    either: (_onSuccess, onFailure) => onFailure(reason),
    mapEither: (_onSuccess, onFailure) => onFailure(reason),
})

export const tryCatch = <S, F>(fn: () => S, onError: (error: unknown) => F): Result<S, F> => {
    try { return success(fn()) }
    catch (e) { return failure(onError(e)) }
}
